tests/helpers.rs

```diff
@@ -230,3 +268,184 @@ pub fn stop_ic_gateway(mut process: Child) {
+        let pic = PocketIcBuilder::new().with_nns_subnet().build();
+        pic.auto_progress();
+        info!("pocket-ic server started");
+
+        let root_key = pic.root_key().expect("failed to get root key");
+        let mut root_key_path = env::current_dir().expect("failed to get working directory");
+        root_key_path.push("root_key.der");
+        let mut file = File::create(&root_key_path).expect("failed to create file");
+        file.write_all(&root_key)
+            .expect("failed to write key to file");
+
+        let ic_gateway_addr =
+            SocketAddr::from_str(ic_gateway_addr).expect("failed to parse address");
+        let ic_url = format!("{}instances/{}/", pic.get_server_url(), pic.instance_id());
+        let process = start_ic_gateway(
+            &ic_gateway_addr.to_string(),
+            ic_gateway_domain,
+            &ic_url,
+            root_key_path,
+        );
+
+        Self {
+            ic_gateway_process: process,
+            ic_gateway_addr,
+            ic_gateway_domain: ic_gateway_domain.to_string(),
+            pic,
+        }
+    }
+}
+
+impl Drop for TestEnv {
+    fn drop(&mut self) {
+        stop_ic_gateway(&mut self.ic_gateway_process);
+    }
+}
+
+#[derive(Debug)]
+pub struct ExpectedResponse {
+    pub status: Option<StatusCode>,
+    pub body: Option<Vec<u8>>,
+    pub headers: Option<HashMap<String, String>>,
+}
+
+impl ExpectedResponse {
+    pub fn new(
+        status: Option<StatusCode>,
+        body: Option<Vec<u8>>,
+        headers: Option<HashMap<String, String>>,
+    ) -> Self {
+        Self {
+            status,
+            body,
+            headers,
+        }
+    }
+
+    pub fn with_header(mut self, key: &str, value: &str) -> Self {
+        if let Some(ref mut headers) = self.headers {
+            headers.insert(key.to_string(), value.to_string());
+        } else {
+            self.headers = Some(HashMap::from_iter(vec![(
+                key.to_string(),
+                value.to_string(),
+            )]));
+        }
+        self
+    }
+}
+
+pub async fn check_response(response: Response, expected: &ExpectedResponse) -> anyhow::Result<()> {
+    let status = response.status();
+    let headers = response.headers().clone();
+    let body_bytes = response.bytes().await?.to_vec();
+
+    if let Some(expected_status) = expected.status {
+        if expected_status != status {
+            anyhow::bail!("unexpected status code: got {status}, expected {expected_status}",);
+        }
+    }
+
+    if let Some(ref expected_body) = expected.body {
+        if &body_bytes != expected_body {
+            anyhow::bail!(
+                "unexpected response body: got size={}, expected={}",
+                body_bytes.len(),
+                expected_body.len()
+            );
+        }
+    }
+
+    if let Some(ref expected_headers) = expected.headers {
+        let mut expected_headers_sorted = HashMap::<String, Vec<String>>::new();
```

Comment:
nit: it might be more concise to use `itertools` here (and below) to do smth like:

```
use itertools::Itertools;

let expected_headers_sorted = HashMap::from_iter(expected_headers.iter().map{|k, v| {
 (
  k.trim().to_ascii_lowercase(),
  v.split(',').map(|x| x.trim().to_ascii_lowercase()).sorted().collect::<Vec<_>>()
 )
});
```

not that it matters much